plutus-core/plutus-core/src/PlutusCore/Builtin/Runtime.hs

```diff
@@ -77,13 +82,28 @@ toBuiltinRuntime :: cost -> BuiltinMeaning val cost -> BuiltinRuntime val
 toBuiltinRuntime cost (BuiltinMeaning sch f exF) =
     BuiltinRuntime (typeSchemeToRuntimeScheme sch) f (exF cost)
 
--- | Calculate runtime info for all built-in functions given denotations of builtins
--- and a cost model.
-toBuiltinsRuntime
-    :: (cost ~ CostingPart uni fun, HasConstantIn uni val, ToBuiltinMeaning uni fun)
-    => cost -> BuiltinsRuntime fun val
-toBuiltinsRuntime cost =
-    BuiltinsRuntime . tabulateArray $ toBuiltinRuntime cost . toBuiltinMeaning
+-- | This is a function turned into a class for the sake of specialization and inspection.
+-- Every instance of this class must use the default implementation of 'toBuiltinsRuntime',
+-- i.e. not define 'toBuiltinsRuntime' at all. Having this class allows us to inlining away lots of
+-- abstractions used by the builtins machinery.
+--
+-- For performance-critical code both @fun@ and @val@ need to be specified as concrete data types
+-- for inlining to happen. In other places it's fine to only specify one of them.
+class ToBuiltinsRuntime fun val where
```

Comment:
Let me try and write down what I think this is doing.

We want to specialize `toBuiltinsRuntime` to specific choices of `fun` and `val`. It's not enough to inline `toBuiltinsRuntime`, because then we will only optimize the generic generated code at the call site, rather than generating specialized code (is this true?). Requiring a type class instance for each combination of `fun` and `val` creates a definition site which can be individually optimized, and then inlined.

That suggests the following question: does this basically amount to a `SPECIALIZE` pragma? Note that `SPECIALIZE` pragmas can be given for *imported* functions so long as they're `INLINABLE`.

Reply:
I'm also genuinely quite unclear why inlining `toBuiltinsRuntime` doesn't do the job more simply.

Reply:
Probably also need to inline `toMachineParameters`.

Reply:
> It's not enough to inline `toBuiltinsRuntime`, because then we will only optimize the generic generated code at the call site, rather than generating specialized code (is this true?).

Yes, unless the call site has `uni` and `fun` specialized. If it doesn't, we'll also need to inline the call site of the call site. And so on until we reach a call site with `uni` and `fun` specialized. With the class approach we're just passing around a dictionary with specialized and optimized stuff in it and don't need to worry about chains of call sites, they can do whatever they want.

Note that we're running the CEK machine from different places, so we'd need to track all of them. How would we even do that? `inspection-testing`? Let's-occasionally-dump-Core-and-check-if-it's-fine-at-every-call-site?

Using a class just seems more reliable than hoping a `SPECIALIZE` pragma will fire or a chain of calls won't have a missing `INLINE` somewhere in the middle. And `INLINE` is sometimes not enough, an explicit call to `inline` is needed and all of that can change from one GHC version to another. Having a single place where we need to worry about checking if things still get optimized properly is much less of a headache.

Reply:
Well, except this is only for the call to `toBuiltinsRuntime`. Which isn't called in that many places. From a quick grep, about 2. 

> Note that we're running the CEK machine from different places, so we'd need to track all of them. How would we even do that?

This is why we have benchmarking, no? We've done all kinds of things that we might accidentally break.

Reply:
> Which isn't called in that many places. From a quick grep, about 2.

4, 2 of which are for tests. 1 of the remaining 2 is `toMachineParameters` general over `uni` and `fun`, so we need to track down all of its usages too, of which we have, em, 6? I'd rather not perform this exercise ever again.

> This is why we have benchmarking, no?

Benchmarking is an entirely separate call site. What if it gets inlined there, but not in one of the ledger API?

Anyways, I don't really care. If you want me to make it go via an explicit `SPECIALIZE` pragma, I'll do that. But I'm gonna  refuse to track down all the call sites and check that stuff gets inlined properly at each of them. I mean, inlining literally breaks apart when I remove a goddamn equality constraint (I checked that multiple times), I'm not going to commit to ensuring that GHC doesn't do something extremely weird at one of the call sites, because there's a type synonym in there or whatever.

Reply:
Well, we now have `validation-full` that goes via the actual entry point, so we should be able to check all the things we care about.

This version just also seems a bit fragile :/ I'm not sure.

Reply:
> This version just also seems a bit fragile :/ I'm not sure.

Do you mean that it's ensuring that we use the optimized version of `toBuiltinsRuntime` via a class is fragile or that we may not have all the optimizations in there that we want to have? Because the former seems pretty reliable, especially compared to using `SPECIALIZE`, and the latter is equally unreliable regardless of whether we use a class or a `SPECIALIZE` pragma or just a call to `inline`, but with the class we at least only have a single place to look into and we ensure that this place does not have any unrelated code (2.5k of Core is already a lot to digest).

I don't think we can do anything about the unreliability of the GHC's optimizer. We could try using `inspection-testing` in strategic places (whatever they might be), but I think there's no point about worrying it right now. Let's fix the performance first and then we can think about preserving it.

> Well, we now have `validation-full` that goes via the actual entry point, so we should be able to check all the things we care about.

A 3-8% [speedup](https://github.com/input-output-hk/plutus/pull/4418#issuecomment-1045998794) of evaluation corresponds to only a 1-3% [speedup](https://github.com/input-output-hk/plutus/pull/4418#issuecomment-1046022587) of full validation, which is really close to the noise threshold. I'd really like going in the opposite direction and testing the performance of the builtin machinery via a dedicated benchmark instead of shoving everything into a single huge run and then trying to separate actual changes from noise.

Reply:
> I'd really like going in the opposite direction and testing the performance of the builtin machinery via a dedicated benchmark instead of shoving everything into a single huge run and then trying to separate actual changes from noise.

Yes, although the other way to look at that is that we shouldn't necessarily spend as much time trying to optimize the builtin machinery if the overall effect is small!

Reply:
I want to have a little try and see if I can make a version with SPECIALIZE work also, just for comparison.

Reply:
> I'd really like going in the opposite direction and testing the performance of the builtin machinery via a dedicated benchmark instead of shoving everything into a single huge run and then trying to separate actual changes from noise.

We kind of have that in the cost model benchmarks for the builtins, if you're willing to wait 12 hours or so.  I'm not suggesting that seriously because (a) it takes too long, and (b) it tells us far more than we need to know.  Maybe we could do something like extracting some small subset for checking the performance of the machinery though?  I'm already trying to make some attempt to do that with some extra `Nop` benchmarks that are supposed to give some measure of the overhead of calling a builtin so that we can just model the cost of running the denotation.   

I suppose that an advantage of having some separate benchmarks would be that it'd make it easier to spot regressions in the efficiency of the builtin machinery.  That makes me a bit nervous because a lot of it depends on delicate implementation choices and I can imagine things going backwards if something changes in GHC.

Reply:
> A speedup of 31.22% seems to be quite a decent one.

I can't of course just add up those individual changes, I'm cutting some corners here.

Reply:
> Add to that the monomorphic makeKnown PR giving us another 4.87% and the inlined toBuiltinsRuntime PR giving us another 4.03%

My rusty math skills suggest that with these two PRs we'll get a total of 34% speedup, which I think is not bad at all.